lib/connections/limiter.go

```diff
@@ -124,6 +127,87 @@ func (lim *limiter) VerifyConfiguration(from, to config.Configuration) error {
 	return nil
 }
 
+func nextChange(to config.Configuration) time.Duration {
+	now := time.Now()
+
+	beginHour := to.Options.ScheduledRatesFromHour
+	beginMinute := to.Options.ScheduledRatesFromMinute
+	begin := time.Date(now.Year(), now.Month(), now.Day(), beginHour, beginMinute, 0, 0, now.Location())
+
+	tillBegin := begin.Sub(now)
+
+	var waitTime time.Duration
+
+	if tillBegin >= 0 {
+		waitTime = tillBegin
+	} else {
+		endHour := to.Options.ScheduledRatesToHour
+		endMinute := to.Options.ScheduledRatesToMinute
+		end := time.Date(now.Year(), now.Month(), now.Day(), endHour, endMinute, 0, 0, now.Location())
+		if end.Sub(begin) < 0 {
+			end = end.AddDate(0, 0, 1)
+		}
+		tillEnd := end.Sub(now)
+		if tillEnd >= 0 {
+			waitTime = tillEnd
+		} else {
+			begin = begin.AddDate(0, 0, 1)
+			waitTime = begin.Sub(now)
+		}
+	}
+
+	return waitTime
+}
+
+func (lim *limiter) timerLoop(timer *time.Timer, to config.Configuration) {
+	for range timer.C {
+		lim.mu.Lock()
+		defer lim.mu.Unlock()
+
+		now := time.Now()
+
+		beginHour := to.Options.ScheduledRatesFromHour
+		beginMinute := to.Options.ScheduledRatesFromMinute
+		begin := time.Date(now.Year(), now.Month(), now.Day(), beginHour, beginMinute, 0, 0, now.Location())
+
+		endHour := to.Options.ScheduledRatesToHour
+		endMinute := to.Options.ScheduledRatesToMinute
+		end := time.Date(now.Year(), now.Month(), now.Day(), endHour, endMinute, 0, 0, now.Location())
+
+		var sendLimitValue, recvLimitValue int
+
+		if now.Sub(begin) >= 0 && end.Sub(now) >= 0 {
+			sendLimitValue = to.Options.MaxScheduledSendKbps
+			recvLimitValue = to.Options.MaxScheduledRecvKbps
+		} else {
+			sendLimitValue = to.Options.MaxSendKbps
+			recvLimitValue = to.Options.MaxRecvKbps
+		}
+
+		sendLimitStr := "is unlimited"
+		recvLimitStr := "is unlimited"
+
+		if recvLimitValue <= 0 {
+			lim.read.SetLimit(rate.Inf)
+		} else {
+			lim.read.SetLimit(1024 * rate.Limit(recvLimitValue))
+			recvLimitStr = fmt.Sprintf("limit is %d KiB/s", recvLimitValue)
+		}
+
+		if sendLimitValue <= 0 {
+			lim.write.SetLimit(rate.Inf)
+		} else {
+			lim.write.SetLimit(1024 * rate.Limit(sendLimitValue))
+			sendLimitStr = fmt.Sprintf("limit is %d KiB/s", sendLimitValue)
+		}
+
+		l.Infof("Overall send rate %s, receive rate %s", sendLimitStr, recvLimitStr)
```

Comment:
Setting the limits duplicates the existing code in `CommitConfiguration`, should probably become a utility function.

Reply:
Done